### PR TITLE
[feature] Add getMinimumProductVersion

### PR DESCRIPTION
**Pull request by ferozco**
## Before this PR
There was no easy way for users to interact with their product dependencies at configuration time 

## After this PR
We expose new `getMinimumProductVersion` method

@robert3005 for SA

### CODE REVIEW

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.java [hunk 1]
```diff
@@ -77,6 +77,7 @@
             .setSerializationInclusion(JsonInclude.Include.NON_NULL)
             .setPropertyNamingStrategy(PropertyNamingStrategy.KEBAB_CASE)
             .enable(SerializationFeature.INDENT_OUTPUT);
+    public static final String PRODUCT_DEPENDENCIES_LOCK = "product-dependencies.lock";
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.java]: Might as well move this constant to `ProductDependenciesLockFile.java`

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependencyLockFile.java [hunk 2]
```diff
@@ -16,13 +16,39 @@
...
+
+    public static List<ProductDependency> fromString(String contents, String projectVersion) {
+        return Splitter.on("\n").splitToList(contents).stream()
+                .flatMap(line -> {
+                    Matcher matcher = LOCK_PATTERN.matcher(line);
+                    if (matcher.matches()) {
+                        return Stream.of(new ProductDependency(
+                                matcher.group("group"),
+                                matcher.group("name"),
+                                matcher.group("min").equals(PROJECT_VERSION) ? projectVersion : matcher.group("min"),
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependencyLockFile.java:45]: can we have a unit test that covers this `$productVersion` substitution logic pls?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionPlugin.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import java.io.File;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Optional;
+import org.gradle.api.GradleException;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.util.GFileUtils;
+
+public final class BaseDistributionPlugin implements Plugin<Project> {
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionPlugin.java]: Base is possibly a bit misleading here, as this literally only adds a single `getMinimumProductVersion` utility... can we try some alternate names?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependencyLockFile.java [hunk 4]
```diff
@@ -16,13 +16,39 @@
...
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public final class ProductDependencyLockFile {
 
     private static final String HEADER = "# Run ./gradlew --write-locks to regenerate this file\n";
+    private static final String PROJECT_VERSION = "$projectVersion";
+    private static final Pattern LOCK_PATTERN = Pattern.compile(
+            "(?<group>[^:]+):(?<name>[^ ]+) \\((?<min>[^,]+), (?<max>[^\\)]+)\\)");
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependencyLockFile.java]: Worth a ^ and $ to be strict against any weird modifications?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependencyLockFile.java [hunk 5]
```diff
@@ -16,13 +16,39 @@
...
+
+    public static List<ProductDependency> fromString(String contents, String projectVersion) {
+        return Splitter.on("\n").splitToList(contents).stream()
+                .flatMap(line -> {
+                    Matcher matcher = LOCK_PATTERN.matcher(line);
+                    if (matcher.matches()) {
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependencyLockFile.java:41]: :+1: neat way to exclude comments!

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/GradleIntegrationSpec.groovy [hunk 6]
```diff
@@ -24,6 +24,7 @@ class GradleIntegrationSpec extends IntegrationTestKitSpec {
 
     def setup() {
         keepFiles = true
+        debug = true
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/GradleIntegrationSpec.groovy]: safe to revert this now?

**Review by iamdanfox — APPROVED** [review]
Looks solid, just cosmetic requests.  We should probably mention this in the README too otherwise people will slack us forever

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/GetMinimumVersionTest.groovy [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import nebula.test.ProjectSpec
+import org.gradle.api.GradleException
+import org.gradle.api.Project
+import org.gradle.testfixtures.ProjectBuilder
+
+class GetMinimumVersionTest extends ProjectSpec {
+    Project project
+
+    def setup() {
+        project = ProjectBuilder.builder().build()
```
Comment (dansanduleac) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/GetMinimumVersionTest.groovy]: ProjectSpec gives you this

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/GetMinimumVersionTest.groovy [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        then:
+        result == "1.0.0"
+    }
+
+    def "resolves project versions"() {
+        project.version = "1.0.0"
+        project.file(ProductDependencyLockFile.LOCK_FILE) << ProductDependencyLockFile.asString(
+                [new ProductDependency("com.palantir.product", "test", "1.0.0", "1.x.x", null)],
+                [new ProductId("com.palantir.other", "test")] as Set<ProductId>,
+                "1.0.0"
```
Comment (dansanduleac) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/GetMinimumVersionTest.groovy]: > this test isn't testing anything
@ferozco 2019

**Review by dansanduleac — APPROVED** [review]

**Pull request merged**